src/tactic/core.lean

```diff
@@ -353,6 +344,62 @@ do to_remove ← hs.mfilter $ λ h, do {
   to_remove.mmap' (λ h, try (clear h)),
   return (¬ to_remove.empty ∨ goal_simplified)
 
+/-- `revert_after e` reverts all local constants after local constant `e`. -/
+meta def revert_after (e : expr) : tactic ℕ := do
+  l ← local_context,
+  [pos] ← return $ l.indexes_of e | pp e >>= λ s, fail format!"No such local constant {s}",
+  let l := l.drop pos.succ, -- all local hypotheses after `e`
+  revert_lst l
+
+/-- `generalize' e n` is similar as `generalize` but also succeeds when `e` does not occur in the
+  goal, in which case it just calls `assert`. It already introduces the generalized variable. -/
+meta def generalize' (e : expr) (n : name) : tactic expr :=
+(generalize e n >> intro1) <|> note n none e
+
+/- Various tactics related to local definitions (local constants of the form `x : α := t`).
+  We call `t` the body or value of `x`. -/
+
+/-- `local_def_value e` returns the value of the expression `e`, assuming that `e` has been defined
+locally using a `let` expression. Otherwise it fails. -/
+meta def local_def_value (e : expr) : tactic expr :=
```

Comment:
Would this and the following be better implemented in C++? IIRC local definitions are stored in the type context, right? @EdAyers exposed some of the type context to the tactic monad in 3.5c. Rather than this kind of hackish way to get the local def value, maybe there should be a function to query the type context directly.

Reply:
Yes, it would almost certainly be better to implement this in C++. 
I don't know how to do that, though.

Reply:
I think this is already implemented:
```lean
tactic.unsafe.type_context.run $ do
lctx <- get_local_context,
some ldecl <- pure (lctx.get_local_decl e.local_uniq_name),
some let_val <- pure ldecl.value,
pure let_val
```

Reply:
Nice! Do you know if it possible to run `tactic.pp` inside the `type_context` monad (for error message handling)?

Reply:
I didn't find anything, and I don't think there is a way to run tactics from the type context monad.